File: main/src/ca/uwaterloo/flix/runtime/solver/api/LatticeOps.scala
package ca.uwaterloo.flix.runtime.solver.api

import flix.runtime.ProxyObject

// TODO: Java interface

// TODO: Need to standardize on a functional interface for all functions... Perhaps Function[AnyRef, AnyRef]?

trait LatticeOps {

  def bot: ProxyObject

  def equ: java.util.function.Function[Array[AnyRef], ProxyObject]

  def leq: java.util.function.Function[Array[AnyRef], ProxyObject]

  def lub: java.util.function.Function[Array[AnyRef], ProxyObject]

  def glb: java.util.function.Function[Array[AnyRef], ProxyObject]

}

File: main/src/ca/uwaterloo/flix/runtime/solver/api/LatticeOps.java
package ca.uwaterloo.flix.runtime.solver.api;

import flix.runtime.ProxyObject;

import java.util.function.Function;

public interface LatticeOps {

    ProxyObject bot();

    Function<Object[], ProxyObject> equ();

    Function<Object[], ProxyObject> leq();

    Function<Object[], ProxyObject> lub();

    Function<Object[], ProxyObject> glb();

}
